### ISSUE TITLE
Optional Outputs

### ISSUE THREAD

**Issue by tmdefreitas** [openwdl/wdl#20]
Broad's GDAC often deals with sparsely and inconsistently populated data, and many analyses support a range of acceptable inputs. For example, our methylation preprocessor will correlate methylation data to expression data if it is available for a sample set. However, this means that in addition to optional inputs (already supported by WDL), output files are also optional, or at least conditional upon inputs. Currently cromwell raises an error when an expected output is not present, and I don't believe there is syntax to support it in WDL.

It would be nice to be able to specify that an output is optional. One solution could be to mirror to the optional input syntax, and I could specify optional outputs like this:

```
task my_task {
    ... inputs, commands, etc...

    output {
        required_output="file1"
        optional_output?="file2"
    }
}
```

Having this would eliminate the need to write additional logic to create/handle blank files.

**Comment by scottfrazer** [comment]
Seems legit to me

**Comment by cjllanwarne** [comment]
I think you can get *almost* the same thing already using globs - I believe optional_output = glob("file2") will give you either a 0 or 1 element array of files depending on the existence or otherwise of the file.

**Comment by tmdefreitas** [comment]
Thanks for pointing out that workaround, that would at least allow this:

```
output {
  Array[File] optional_output = glob("file2")
}
```

It would still be nice at some point to not worry about 0/1-element arrays, since that syntax doesn't make it clear that the array will only be size 0 or 1.

**Comment by davidbenjamin** [comment]
I'm running into the same issue writing scripts for Mutect 2.

**Comment by davidbenjamin** [comment]
In my case, it's not that a task may or may not produce an output, it's that some tasks may not be run at all depending on a conditional.

**Comment by vdauwera** [comment]
@kshakir Has this been addressed on the Cromwell side? I remember some chatter but I have no details...

**Comment by kshakir** [comment]
Not yet. Here is the cromwell issue: https://github.com/broadinstitute/cromwell/issues/1838

**Comment by antonkulaga** [comment]
I totally support this issue. We need Option-s to work

**Comment by jtratner** [comment]
See https://github.com/openwdl/wdl/pull/310 for PR adding this to spec (and note that cromwell seems to already support this behavior)

**Comment by hisplan** [comment]
Has this been implemented yet?

**Comment by jtratner** [comment]
It works as described in both dxWDL and in Cromwell.

**Comment by jtratner** [comment]
If there’s a conformance test spec I’d be happy to add to it (or could take
advantage of the tests @orodeh wrote for dxWDL)

Also would not hurt for someone to double check I got the semantics correct.

**Comment by hisplan** [comment]
Right now I'm using the `glob` trick. I just tried `File?` with WDL 1.0 + Cromwell 45.1 (latest as of now), but it failed with the following message:

```
failed. Job exited without an error, exit code 0. PAPI error code 5. 10: Failed to delocalize files: failed to copy the following files
```

Is there anything that I need to use/enable in order to use `File? optionalOut`?

**Comment by jdidion** [comment]
Addressed in v1.1
